### DISCUSSION TITLE
Use datatype instances as macro parameter editors

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#14622] [Features and ideas]
Currently macro parameter editors are data editors marked as `MacroParameter`: https://github.com/umbraco/Umbraco-CMS/pull/14594

By default it use the default configuration of the datatype or if a special configuration is needed, it need to implement a macro parameter editor with the specific configuration.

I wonder if macro parameters editors instead could be fetched from existing datatype instances (which is supported as macro parameter). That way if may be simpler to configure single/multiple pickers to be used in macro properties..

![image](https://github.com/umbraco/Umbraco-CMS/assets/2919859/3a880a02-56cc-438d-82ad-fd9e947f3bf1)

... and the datatype instances could possible show the macro references where it is used.

**Example**

![image](https://github.com/umbraco/Umbraco-CMS/assets/2919859/e4818a5c-7b96-40d9-b0d3-b0e6a1fe60d0)

**Comment by bergmania** [comment]
We don't have any plans to make changes to macros, [instead we are looking for ways to deprecate them](https://github.com/umbraco/rfcs/blob/0026-blocks-in-the-rte/cms/0026-blocks-in-the-rte.md)

**Reply by bjarnef** [reply]
@bergmania thanks for highlighting this RFC 👍
I think it makes sense to focus on reusable block including a way to embed some of these into richtext editor using element types. Although not called macro, the functionality may be similar as macro - a placeholder replaced with actually rendered content.
But this would allow use to use more complex/custom configuration of "macros" (or blocks) in RTE and re-use existing datatype configuration, where macros currently need a parameter instance per configuration (e.g. single or multi picker).

We don't use macros a lot, but sometimes to insert forms (but mostly I prefer a block using a form picker). Sometimes also used to insert a login form, but mostly I prefer a specific page type an render to macro or component directly in the template to avoid content editors removing it by mistake.

**Reply by jdpnielsen** [reply]
I would love an implementation like Sanity's Portable text as a replacement for Macros: https://github.com/portabletext/portabletext

**Reply by bergmania** [reply]
@jdpnielsen we already have someting similar for the delivery api, if you configure it to use [RichTextOutputAsJson](https://docs.umbraco.com/umbraco-cms/reference/content-delivery-api)

**Reply by jdpnielsen** [reply]
Nice!
